config/proposer/loader.go

```diff
@@ -0,0 +1,281 @@
+	for _, method := range psl.loadMethods {
+		switch method {
+		case defaultFlag:
+			suggestedFee := cliCtx.String(flags.SuggestedFeeRecipientFlag.Name)
+			if !common.IsHexAddress(suggestedFee) {
+				return nil, errors.New("default fee recipient is not a valid Ethereum address")
+			}
+			if err := config.WarnNonChecksummedAddress(suggestedFee); err != nil {
+				return nil, err
+			}
+			defaultConfig := &validatorpb.ProposerOptionPayload{
+				FeeRecipient: suggestedFee,
+			}
+			if psl.options.builderConfig != nil {
+				defaultConfig.Builder = psl.options.builderConfig.ToConsensus()
+			}
+			if loadConfig == nil {
+				loadConfig = &validatorpb.ProposerSettingsPayload{}
+			}
+			loadConfig.DefaultConfig = defaultConfig
+		case fileFlag:
+			var settingFromFile *validatorpb.ProposerSettingsPayload
+			if err := config.UnmarshalFromFile(cliCtx.Context, cliCtx.String(flags.ProposerSettingsFlag.Name), &settingFromFile); err != nil {
+				return nil, err
+			}
+			if settingFromFile == nil {
+				return nil, errors.New("proposer settings is empty after unmarshalling from file")
+			}
+			loadConfig = psl.processProposerSettings(settingFromFile, loadConfig)
+		case urlFlag:
+			var settingFromURL *validatorpb.ProposerSettingsPayload
+			if err := config.UnmarshalFromURL(cliCtx.Context, cliCtx.String(flags.ProposerSettingsURLFlag.Name), &settingFromURL); err != nil {
+				return nil, err
+			}
+			if settingFromURL == nil {
+				return nil, errors.New("proposer settings is empty after unmarshalling from url")
+			}
+			loadConfig = psl.processProposerSettings(settingFromURL, loadConfig)
+		case onlyDB:
+			loadConfig = psl.processProposerSettings(nil, loadConfig)
+		case none:
+			if psl.options.builderConfig != nil {
+				// if there are no proposer settings provided, create a default where fee recipient is not populated, this will be skipped for validator registration on validators that don't have a fee recipient set.
+				// skip saving to DB if only builder settings are provided until a trigger like keymanager API updates with fee recipient values
+				option := &validatorService.ProposerOption{
+					BuilderConfig: psl.options.builderConfig.Clone(),
+				}
+				if loadConfig == nil {
+					loadConfig = &validatorpb.ProposerSettingsPayload{}
+				}
+				loadConfig.DefaultConfig = option.ToConsensus()
+			}
+		default:
+			return nil, errors.New("load method for proposer settings does not exist")
+		}
+	}
+
+	// exit early if nothing is provided
+	if loadConfig == nil {
+		log.Warn("No proposer settings were provided")
+		return nil, nil
+	}
+	ps, err := validatorService.ProposerSettingFromConsensus(loadConfig)
+	if err != nil {
+		return nil, err
+	}
+	if err := psl.db.SaveProposerSettings(cliCtx.Context, ps); err != nil {
+		return nil, err
+	}
+	return ps, nil
+}
+
+func (psl *SettingsLoader) processProposerSettings(loadedSettings, dbSettings *validatorpb.ProposerSettingsPayload) *validatorpb.ProposerSettingsPayload {
```

Comment:
The name `dbSettings` gives the impression that these settings are coming from the `db`, which is not (always) the case.

Reply:
To be honest I feel quite confused with this function.

My understanding is that this function will globally override `loadedSettings` with `dbSettings`. (`loadedSettings` will be mutated, but also return by the function.)
In some condition, `loadedSettings` will also depend on `psl`. 

After spending some times on this function, I'm still not clear on exactly why things are done.